SeaPkg/Library/BasePeCoffValidationLib/Validation.c

```diff
@@ -0,0 +1,325 @@
+  - MemAttr: Validates that the specified region in the target image buffer belongs to a user page that is mapped inside MM
+             and the page attributes match the requirements specified by the validation entry.
+  - SelfRef: Validates that the specified region in the target image buffer matches the content in the original image buffer as
+             specified by TargetOffset in the validation entry.
+  - Pointer: Validates that the specified region in the target image buffer is a pointer, and that the pointer is not NULL.
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include <Uefi.h>
+#include <Base.h>
+#include <SeaAuxiliary.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/PeCoffValidationLib.h>
+#include <Library/SafeIntLib.h>
+
+EFI_STATUS
+EFIAPI
+GetMemoryAttributes (
+  IN  EFI_PHYSICAL_ADDRESS  PageTableBase,
+  IN  EFI_PHYSICAL_ADDRESS  BaseAddress,
+  IN  UINT64                Length,
+  OUT UINT64                *Attributes
+  );
+
+/**
+  Helper function that will evaluate the page where the input address is located belongs to a
+  user page that is mapped inside MM.
+  @param  Address           Target address to be inspected.
+  @param  Size              Address range to be inspected.
+  @param  IsUserRange       Pointer to hold inspection result, TRUE if the region is in User pages, FALSE if
+                            the page is in supervisor pages. Should not be used if return value is not EFI_SUCCESS.
+  @return     The result of inspection operation.
+**/
+EFI_STATUS
+InspectTargetRangeAttribute (
+  IN  EFI_PHYSICAL_ADDRESS  PageTableBase,
+  IN  EFI_PHYSICAL_ADDRESS  Address,
+  IN  UINTN                 Size,
+  OUT UINT64                *MemAttribute
+  )
+{
+  EFI_STATUS            Status;
+  EFI_PHYSICAL_ADDRESS  AlignedAddress;
+  UINT64                Attributes;
+
+  if ((Address < EFI_PAGE_SIZE) || (Size == 0) || (MemAttribute == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Done;
+  }
+
+  AlignedAddress = ALIGN_VALUE (Address - EFI_PAGE_SIZE + 1, EFI_PAGE_SIZE);
+
+  // To cover head portion from "Address" alignment adjustment
+  Status = SafeUintnAdd (Size, Address - AlignedAddress, &Size);
+  if (EFI_ERROR (Status)) {
+    goto Done;
+  }
+
+  // To cover the tail portion of requested buffer range
+  Status = SafeUintnAdd (Size, EFI_PAGE_SIZE - 1, &Size);
+  if (EFI_ERROR (Status)) {
+    goto Done;
+  }
+
+  Size &= ~(EFI_PAGE_SIZE - 1);
+
+  // Go through page table and grab the entry attribute
+  Status = GetMemoryAttributes (PageTableBase, AlignedAddress, Size, &Attributes);
+  if (!EFI_ERROR (Status)) {
+    *MemAttribute = Attributes;
+    goto Done;
+  }
+
+Done:
+  return Status;
+}
+
+/**
+  Validates a specific region in the target image buffer denoted by [Hdr->Offset: Hdr->Offset + Hdr->Size]
+  is not a zero buffer.
+  @param[in] TargetImage  The pointer to the target image buffer.

```

Comment:
For all of these, the sections need some whitespace separation.

```suggestion
  is not a zero buffer.

  @param[in] TargetImage  The pointer to the target image buffer.
```

Reply:
Fixed the spacing here and in multiple other places!